nixos/modules/system/boot/luksroot.nix

```diff
@@ -277,38 +277,40 @@ let
 
         [ "$opened" == false ] && die "Maximum authentication errors reached"
 
-        echo -n "Gathering entropy for new salt (please enter random keys to generate entropy if this blocks for long)..."
-        for i in $(seq ${toString yubikey.saltLength}); do
-            byte="$(dd if=/dev/random bs=1 count=1 2>/dev/null | rbtohex)";
-            new_salt="$new_salt$byte";
-            echo -n .
-        done;
-        echo "ok"
-
-        new_iterations="$iterations"
-        ${optionalString (yubikey.iterationStep > 0) ''
-        new_iterations="$(($new_iterations + ${toString yubikey.iterationStep}))"
+        ${optionalString yubikey.renewKey ''
+          echo -n "Gathering entropy for new salt (please press some random keys to generate entropy if this blocks for long)..."
+          for i in $(seq ${toString yubikey.saltLength}); do
+              byte="$(dd if=/dev/random bs=1 count=1 2>/dev/null | rbtohex)";
+              new_salt="$new_salt$byte";
+              echo -n .
+          done;
+          echo "ok"
+
+          new_iterations="$iterations"
+          ${optionalString (yubikey.iterationStep > 0) ''
+          new_iterations="$(($new_iterations + ${toString yubikey.iterationStep}))"
+          ''}
+
+          new_challenge="$(echo -n $new_salt | openssl-wrap dgst -binary -sha512 | rbtohex)"
+
+          new_response="$(ykchalresp -${toString yubikey.slot} -x $new_challenge 2>/dev/null)"
+
+          if [ ! -z "$k_user" ]; then
+              new_k_luks="$(echo -n $k_user | pbkdf2-sha512 ${toString yubikey.keyLength} $new_iterations $new_response | rbtohex)"
+          else
+              new_k_luks="$(echo | pbkdf2-sha512 ${toString yubikey.keyLength} $new_iterations $new_response | rbtohex)"
+          fi
+
+          echo -n "$new_k_luks" | hextorb > /crypt-ramfs/new_key
+          echo -n "$k_luks" | hextorb | ${cschange} --key-file=- /crypt-ramfs/new_key
+
+          if [ $? == 0 ]; then
+              echo -ne "$new_salt\n$new_iterations" > /crypt-storage${yubikey.storage.path}
```

Comment:
a `sync /crypt-storage${yubikey.storage.path}` should probably be added after this line.

Reply:
Are you sure it is necessary?

Reply:
Also, this code directly comes from the current module. 
I am not sure this in the code of the PR.

Reply:
I'm not sure, but it would reduce the possibilities for corruption errors.

Reply:
https://github.com/NixOS/nixpkgs/pull/115451